Fix issues with relative templates in includes and incorrect id

dane-harnett opened the pull request:
After trying to use this in a non-trivial way in our rather complex app I found some issues with the include support.

Mainly around relative templates and the ID references that templates are stored within twig.

webberig commented:
Thanks for your PR! 

Until now I've been able to put all my twig views in a single folder which is far from ideal indeed. I realize that the current stable version will probably not work when using relative path require statements. Which you hopefully have solved! :-)

webberig commented on test/extends.test.js:
```diff
@@ -17,7 +17,7 @@ describe("extend", function() {
       result.should.have.type("string");
 
       // verify the generated module imports the `include`d templates
-      result.should.match(/require\(\"twig\!a\.html\.twig\"\);/);
```
I don't think removing the twig loader prefix is a good idea. The webpack user would be forced to use a config file to map twig files to the twig loader, which is never the case in other loaders.

It shouldn't be too hard adding this to the included templates after processing.

dane-harnett replied:
For some reason when keeping the twig loader prefix, the webpack uses a different loader context for those files and doesn't use the query config that I have defined in my webpack config file. Any ideas on how to solve that?

webberig commented:
What query config do you mean ? Can you post an example ?
I currently don't have much time to try this out myself, I need a couple of days...

dane-harnett commented:
I'll try to describe the problem in detail :)

The issue was with the identifier that we were assigning the template to within twig.
In the current release this is just the filename (path excluded), this means if using template within other directories, there could be clashes.

So I changed the ID to be the full path to the template.
I also re-assigned the reference in the template (during compilation) so that it referenced the full path rather than a relative one.

This worked.

However I wanted to also be able to include templates from a designated "templateRoot" directory otherwise all includes would have needed to be relative to the current template.

To do this I added a query option that can be set via webpack (config file or in the require statement) to set the templateRoot, if this option is set then the full path has the templateRoot removed from it when creating the ID allowing the consumer to use both relative and templateRoot relative includes.

However when adding this option, setting the twig loader within the require statement broke.

I'm happy to revert that just as long as we know that in that case it seems that all includes must be relative to the current template (probably not too much of an issue)

webberig commented:
I finally got some time to look at this...

This is my testing setup:
![image](https://cloud.githubusercontent.com/assets/3446164/12374242/87c18188-bc95-11e5-922e-07aac2dad15c.png)

index.js requires `views/a.html.twig` and `views/b.html.twig`.
a.html.twig also includes `./subfolder/b.html.twig`

As expected, I got an error when running webpack: You may need an appropriate loader to handle this file type. So I added the `twig!` loader prefix.

The project is building just fine, but when I execute the built package, twig returns an error: 

    Error parsing twig template E:Projects3-front  est-twig-loaderiewsa.html:
    TypeError: Cannot set property 'options' of null

Just to be sure: is this what you're trying to achieve ?
I'll try to figure out this bug...

webberig commented:
This could help:
http://webpack.github.io/docs/how-to-write-a-loader.html#should-not-embed-absolute-paths

dane-harnett commented:
Hi,

The `Cannot set property 'options' of null` error is due to the twigjs runtime not being able to find the included template as it is looking for a different ID than what the template is stored under.

I'll take a look at that link and see what I can do.

I've also found a limitation of my implementation of not being able to reference templates in higher level relative directories eg: `../template.html.twig`

webberig commented:
I think I managed to add support to relative template paths. This should now work fine with all kinds of relative paths such as `../` and `./`. However, I need to investigate if this still works if you do use webpack configuration such as `resolve.root`

Can you give my `relative-path-support` branch a try and give me some feedback ?

I was having some trouble with all the refactoring you did in your PR, so I basically started over from my `master`

dane-harnett commented:
OK Cool, I'll take a look later tonight :)

dane-harnett commented:
@webberig I've tested out your branch with my project and it seems to be working as expected, nice work! please merge :)

webberig commented:
OK, I'll close this one and continue my work on #5
